[ UI - Dashboards ] Clone dashboard with cells fails to clone cells as well

<!--

Thank you for reporting a bug in InfluxDB. 

* Please ask usage questions on the Influx Community site.
    * https://community.influxdata.com/
* Please add a :+1: or comment on a similar existing bug report instead of opening a new one.
    * https://github.com/influxdata/influxdb/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+is%3Aclosed+sort%3Aupdated-desc+label%3Akind%2Fbug+
* Please check whether the bug can be reproduced with the latest release.
* The fastest way to fix a bug is to open a Pull Request.
    * https://github.com/influxdata/influxdb/pulls

-->

__Steps to reproduce:__
List the minimal actions needed to reproduce the behavior.

1. Create a dashboard and add at least one cell and a note cell 
2. In the dashboards card for the dashboard click clone
3. Confirm clone

__Expected behavior:__
Describe what you expected to happen.

That the dashboard would be cloned and the clone would be opened as a dashboard

__Actual behavior:__
Describe What actually happened.

There is a message "Failed to create dashboard"

Console log shows

```
generatedRoutes.ts:1836 GET 
http://localhost:9999/api/v2/dashboards/052073d899f7b000/cells/052069f3b777b000/view 404 (Not Found)

generatedRoutes.ts:1836 GET 
http://localhost:9999/api/v2/dashboards/052073d899f7b000/cells/05206a0086b7b000/view 404 (Not Found)

thunks.ts:161 Error: An error occurred cloning the dashboard
    at Object.o (index.ts:57)
    at async http:/localhost:9999/2.31ee54dec8.js:1
```

Force page reload shows the clone.  However on inspection the cells are not present.  And reopening the original cell sometimes results in error. 

__N.B.__

Dashboards with no cells defined clone without issue. 

__Screencasts__

![CloneDashboardWithCells01](https://user-images.githubusercontent.com/25881301/72974459-4e55ec80-3dcf-11ea-9de8-a5877b9a3c5e.gif)
![CloneDashboardWithCells02](https://user-images.githubusercontent.com/25881301/72974460-4e55ec80-3dcf-11ea-817d-0c711bffbbe7.gif)

__Environment info:__

* System info: Run `uname -srm` and copy the output here
```
Linux 4.15.0-74-generic x86_64
```
* InfluxDB version: Run `influxd version` and copy the output here
Detected in nightly build 
```
ts=2020-01-23T08:52:51.141307Z lvl=info msg="Welcome to InfluxDB" log_id=0KWPWBHG000 version=nightly commit=f22201f5a build_date=2020-01-23T08:52:51Z
```
* Other relevant environment details: Container runtime, disk info, etc
Browser: __Chrome 79__

__Config:__
Copy any non-default config values here or attach the full config as a gist or file.

Started with end to end testing enabled

```
--e2e-testing=true
```

<!-- The following sections are only required if relevant. -->

__Logs:__
Include snippet of errors in log.

```
generatedRoutes.ts:1836 GET http://localhost:9999/api/v2/dashboards/052073d899f7b000/cells/052069f3b777b000/view 404 (Not Found)
n.request @ generatedRoutes.ts:1836
N @ generatedRoutes.ts:4906
(anonymous) @ index.ts:42
o @ index.ts:41
(anonymous) @ thunks.ts:150
async function (async)
(anonymous) @ thunks.ts:111
(anonymous) @ index.js:8
(anonymous) @ redux.js:449
handleCloneDashboard @ DashboardCard.tsx:122
I.handleClick @ ContextMenuItem.tsx:55
s @ react-dom.production.min.js:14
h @ react-dom.production.min.js:15
(anonymous) @ react-dom.production.min.js:15
x @ react-dom.production.min.js:17
T @ react-dom.production.min.js:18
M @ react-dom.production.min.js:18
E @ react-dom.production.min.js:21
Dn @ react-dom.production.min.js:84
As @ react-dom.production.min.js:250
je @ react-dom.production.min.js:39
Tn @ react-dom.production.min.js:86
(anonymous) @ react-dom.production.min.js:251
t.unstable_runWithPriority @ scheduler.production.min.js:18
Fs @ react-dom.production.min.js:251
Ln @ react-dom.production.min.js:85
D.Object.isExtensible.Object.isExtensible.e.___hb.e.___hb @ honeybadger.js:466
generatedRoutes.ts:1836 GET http://localhost:9999/api/v2/dashboards/052073d899f7b000/cells/05206a0086b7b000/view 404 (Not Found)
n.request @ generatedRoutes.ts:1836
N @ generatedRoutes.ts:4906
(anonymous) @ index.ts:42
o @ index.ts:41
(anonymous) @ thunks.ts:150
async function (async)
(anonymous) @ thunks.ts:111
(anonymous) @ index.js:8
(anonymous) @ redux.js:449
handleCloneDashboard @ DashboardCard.tsx:122
I.handleClick @ ContextMenuItem.tsx:55
s @ react-dom.production.min.js:14
h @ react-dom.production.min.js:15
(anonymous) @ react-dom.production.min.js:15
x @ react-dom.production.min.js:17
T @ react-dom.production.min.js:18
M @ react-dom.production.min.js:18
E @ react-dom.production.min.js:21
Dn @ react-dom.production.min.js:84
As @ react-dom.production.min.js:250
je @ react-dom.production.min.js:39
Tn @ react-dom.production.min.js:86
(anonymous) @ react-dom.production.min.js:251
t.unstable_runWithPriority @ scheduler.production.min.js:18
Fs @ react-dom.production.min.js:251
Ln @ react-dom.production.min.js:85
D.Object.isExtensible.Object.isExtensible.e.___hb.e.___hb @ honeybadger.js:466
thunks.ts:161 Error: An error occurred cloning the dashboard
    at Object.o (index.ts:57)
    at async http:/localhost:9999/2.31ee54dec8.js:1
```
<!--
__Performance:__
Generate profiles with the following commands for bugs related to performance, locking, out of memory (OOM), etc.

```sh
# Commands should be run when the bug is actively happening.
# Note: This command will run for at least 30 seconds.
curl -o profiles.tar.gz "http://localhost:8086/debug/pprof/all?cpu=true"
curl -o vars.txt "http://localhost:8086/debug/vars"
iostat -xd 1 30 > iostat.txt
# Attach the `profiles.tar.gz`, `vars.txt`, and `iostat.txt` output files.
```
-->

--- a/ui/cypress/e2e/dashboardsIndex.test.ts
+++ b/ui/cypress/e2e/dashboardsIndex.test.ts
@@ -142,6 +142,23 @@ describe('Dashboards', () => {
       cy.getByTestID('dashboard-card').should('have.length', 1)
     })
 
+    it('can clone a dashboard', () => {
+      cy.getByTestID('dashboard-card').should('have.length', 2)
+
+      cy.getByTestID('dashboard-card')
+        .getByTestID('clone-dashboard')
+        .first()
+        .click({force: true})
+
+      cy.fixture('routes').then(({orgs}) => {
+        cy.get('@org').then(({id}: Organization) => {
+          cy.visit(`${orgs}/${id}/dashboards`)
+        })
+      })
+
+      cy.getByTestID('dashboard-card').should('have.length', 3)
+    })
+
     it('can edit a dashboards name', () => {
       const newName = 'new 🅱️ashboard'
 

--- a/ui/src/dashboards/actions/thunks.ts
+++ b/ui/src/dashboards/actions/thunks.ts
@@ -126,7 +126,7 @@ export const cloneDashboard = (
     )
 
     const dash: Dashboard = entities.dashboards[result]
-    const cells = Object.values<Cell>(entities.cells)
+    const cells = Object.values<Cell>(entities.cells || {})
 
     const postResp = await api.postDashboard({
       data: {

--- a/ui/src/dashboards/actions/thunks.ts
+++ b/ui/src/dashboards/actions/thunks.ts
@@ -54,6 +54,7 @@ import {
   Dashboard,
   GetState,
   View,
+  Cell,
   DashboardTemplate,
   Label,
   RemoteDataState,
@@ -110,7 +111,10 @@ export const cloneDashboard = (
     const allDashboardNames = dashboards.map(d => d.name)
     const clonedName = incrementCloneName(allDashboardNames, dashboardName)
 
-    const getResp = await api.getDashboard({dashboardID})
+    const getResp = await api.getDashboard({
+      dashboardID,
+      query: {include: 'properties'},
+    })
 
     if (getResp.status !== 200) {
       throw new Error(getResp.data.message)
@@ -122,7 +126,7 @@ export const cloneDashboard = (
     )
 
     const dash: Dashboard = entities.dashboards[result]
-    const cells = dash.cells.map(cellID => state.resources.cells.byID[cellID])
+    const cells = Object.values<Cell>(entities.cells)
 
     const postResp = await api.postDashboard({
       data: {
@@ -136,7 +140,7 @@ export const cloneDashboard = (
       throw new Error(postResp.data.message)
     }
 
-    const pendingLabels = dash.labels.map(l =>
+    const pendingLabels = getResp.data.labels.map(l =>
       api.postDashboardsLabel({
         dashboardID: postResp.data.id,
         data: {labelID: l.id},
@@ -149,7 +153,13 @@ export const cloneDashboard = (
       throw new Error('An error occurred cloning the labels for this dashboard')
     }
 
-    const clonedViews = await dashAPI.cloneUtilFunc(cells, postResp.data.id)
+    const clonedDashboardID = postResp.data.id
+
+    const clonedViews = await dashAPI.cloneUtilFunc(
+      cells,
+      dashboardID,
+      clonedDashboardID
+    )
 
     const newViews = await Promise.all(clonedViews)
 

--- a/ui/src/dashboards/apis/index.ts
+++ b/ui/src/dashboards/apis/index.ts
@@ -42,11 +42,15 @@ export const updateView = async (
   return viewWithIDs
 }
 
-export const cloneUtilFunc = async (cells: Cell[], dashboardID: string) => {
+export const cloneUtilFunc = async (
+  cells: Cell[],
+  oldDashID: string,
+  clonedDashID: string
+) => {
   const pendingViews = cells.map(cell =>
     api
       .getDashboardsCellsView({
-        dashboardID,
+        dashboardID: oldDashID,
         cellID: cell.id,
       })
       .then(res => {
@@ -66,9 +70,9 @@ export const cloneUtilFunc = async (cells: Cell[], dashboardID: string) => {
     const view = v.data as View
     const cell = cells.find(c => c.id === view.id)
 
-    if (cell && dashboardID) {
+    if (cell) {
       const newCell = await api.postDashboardsCell({
-        dashboardID,
+        dashboardID: clonedDashID,
         data: cell,
       })
 
@@ -77,7 +81,7 @@ export const cloneUtilFunc = async (cells: Cell[], dashboardID: string) => {
       }
 
       return api.patchDashboardsCellsView({
-        dashboardID,
+        dashboardID: clonedDashID,
         cellID: newCell.data.id,
         data: view,
       })

--- a/ui/src/dashboards/components/dashboard_index/DashboardCard.tsx
+++ b/ui/src/dashboards/components/dashboard_index/DashboardCard.tsx
@@ -132,7 +132,11 @@ class DashboardCard extends PureComponent<Props> {
           icon={IconFont.Duplicate}
           color={ComponentColor.Secondary}
         >
-          <Context.Item label="Clone" action={this.handleCloneDashboard} />
+          <Context.Item
+            label="Clone"
+            action={this.handleCloneDashboard}
+            testID="clone-dashboard"
+          />
         </Context.Menu>
         <Context.Menu
           icon={IconFont.Trash}

